src/libsync/propagatedownload.cpp

```diff
@@ -821,8 +827,26 @@ void PropagateDownloadFile::slotGetFinished()
     validator->start(_tmpFile.fileName(), checksumHeader);
 }
 
-void PropagateDownloadFile::slotChecksumFail(const QString &errMsg)
+void PropagateDownloadFile::slotChecksumFail(const QString &errMsg, const QByteArray &checksumType, const QByteArray &checksum, const QString &filePath)
 {
+    if (!checksumType.isEmpty() && !checksum.isEmpty() && !filePath.isEmpty()) {
+        ConfigFile cfgFile;
+
+        if (cfgFile.allowChecksumValidationFail()) {
+            const QString key = checksum + _item->_fileId;
+            const auto numChecksumMismatchEntries = propagator()->_journal->keyValueStoreGetInt(QString(key), 0);
+            if (numChecksumMismatchEntries < numChecksumFailuresAllowed) {
+                qCWarning(lcPropagateDownload) << "Checksum validation has failed" << numChecksumMismatchEntries << " times, but, allowChecksumValidationFail is set.Let's give it another try...";
```

Comment:
Better say that you will retry the download after `numChecksumMismatchEntries` failed attempt due to checksum mismatch. I would put it info level because that indicates a decision but not necessarily something the user has to care about.

Reply:
@mgallien This scenario is actually a warning, because, checksum should not have mismatch under normal conditions, so, clearly, something went wrong. In case we ever face other issue because of allowing such mismatched checksum to proceed with the download, I'd like to be able to spot this warning early. And, yes, I agree with you that the user should not care about it, that's why I've decided not to trigger a GUI warhing just yet, but, rather log this into a text file.